app/src/main/java/org/wikipedia/diff/ArticleEditDetailsFragment.kt

```diff
@@ -0,0 +1,413 @@
+    }
+
+    private fun showThankDialog() {
+        val parent = FrameLayout(requireContext())
+        val dialog: AlertDialog = AlertDialog.Builder(activity)
+                .setView(parent)
+                .setPositiveButton(R.string.thank_dialog_positive_button_text) { _, _ ->
+                    sendThanks()
+                }
+                .setNegativeButton(R.string.thank_dialog_negative_button_text, null)
+                .create()
+        dialog.layoutInflater.inflate(R.layout.view_thank_dialog, parent)
+        dialog.setOnShowListener {
+            dialog.getButton(AlertDialog.BUTTON_NEGATIVE)
+                    .setTextColor(ResourceUtil.getThemedColor(requireContext(), R.attr.secondary_text_color))
+        }
+        dialog.show()
+    }
+
+    private fun sendThanks() {
+        ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).csrfToken
+                .subscribeOn(Schedulers.io())
+                .flatMap {
+                    ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).postThanksToRevision(revisionId, it.query()!!.csrfToken()!!)
+                }
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({
+                    FeedbackUtil.showMessage(activity, getString(R.string.thank_success_message, username))
+                    setButtonTextAndIconColor(thankButton, ResourceUtil.getThemedColor(requireContext(),
+                            R.attr.material_theme_de_emphasised_color))
+                    thankButton.isClickable = false
+                })
+                { Consumer { t: Throwable? -> L.e(t) } }
+    }
+
+    private fun fetchDiffText() {
+        disposables.add(ServiceFactory.getCoreRest(WikiSite.forLanguageCode(languageCode)).getDiff(olderRevisionId, revisionId)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({
+                    displayDiff(it.diffs)
+                }) { t: Throwable? -> L.e(t) })
+    }
+
+    private fun displayDiff(diffs: MutableList<DiffItem>) {
+        for (diff in diffs) {
+            val prefixLength = diffText.text.length
+            val spannableString = SpannableStringBuilder(diffText.text).append(if (diff.text.isNotEmpty()) diff.text else "\n")
+            when (diff.type) {
+                DIFF_TYPE_LINE_WITH_SAME_CONTENT -> {
+                }
+                DIFF_TYPE_LINE_ADDED -> {
+                    updateDiffTextDecor(spannableString, true, prefixLength, prefixLength + diff.text.length)
+                }
+                DIFF_TYPE_LINE_REMOVED -> {
+                    updateDiffTextDecor(spannableString, false, prefixLength, prefixLength + diff.text.length)
+                }
+                DIFF_TYPE_LINE_WITH_DIFF -> {
+                    for (highlightRange in diff.highlightRanges) {
+                        val highlightRangeStart = if (languageCode == "en") highlightRange.start
```

Comment:
Why are there special cases for English here?

Reply:
English characters are 1 byte each.. So can we skip the processing for it?

Reply:
So... English articles can't contain Unicode characters?
https://en.wikipedia.org/wiki/Hindi